### PR TITLE
Persist data

### PR DESCRIPTION
**Pull request by ostafen**
Hi, all! I open this as a draft PR because some points need to be refined and improved after discussion, since this is a feature involving several changes.
The current implementation uses a block file for each table, where blocks are appended in a log-like fashion. 
This should be better than storing a single table block on a separate file, because we avoid the cost of creating/opening file each time.
There is another point I had to address: when syncing a block to disk, there could be some parts in the block which have not been already committed (by increasing the watermark). It would have been natural to wait for all the txns of the block to be aligned with the watermark to start writing the block to disk, but I decided to follow another approach. 
I used a `pendingWritersWg` WaitGroup inside the `TableBlock` object to track the number of goroutines trying to perform an `Insert()` operation on the block.
Before writing the block to disk we wait on the following wait groups:
```go
block.pendingWritersWg.Wait()
block.wg.Wait()
```
The first wait ensures that all goroutines ended the execution of the `Insert()` method, so that we can consequently wait for all compactions to finish on the second WaitGroup, because we are sure no one will trigger other compaction operations for that block.
At this point, the block will no more get modified and, even if there are pending write transactions, all of them should be committed successfully (it is just a matter of waiting for the watermark to be incremented). So, instead of waiting, I implemented the possibility to iterate on all parts, even those which are not aligned with the watermarks, in order to speed up disk writing.

Additional points I think should be addressed:
- I think we should add the possibility to enable/disable persistence when creating a `TableConfig`, and maybe, we could decide to rotate the `BlockFile` when a configurable max size is reached for a file.

- Because now, we store data on disk, the user may also want to configure a root directory where to place all block files.

- We should add a `Close()` method to the `Table` object, in order to do all the necessary clean-up operations when closing a table (wait for all pending blocks to be correctly written to disk and close the current active block file).

- Moreover, when testing, we should also allow the `ActiveMemorySize` to be modified. Actually the limit is to high (512MB), so no disk operation will be triggered.

### CODE REVIEW

**Review by thorfour — COMMENTED** [review]
This looks really good so far after a quick pass. One question around file format

File: granule.go
Comment (ostafen) [reply granule.go:222]: Yes, we can definitely get rid of the flag

File: block_file.go
Comment (thorfour) [reply block_file.go:12]: Each page should have it's own checksum already in the Parquet file format right? The timestamp problem is interesting...

Comment (thorfour) [reply block_file.go:12]: I'm not sure if there's a good way for us to perform this in append mode while keeping the parquet file format which allows us to use the tool/oss ecosystem to inspect etc. I was looking at writing the table block as an appended row group in the file but I think there's some serious drawbacks to that approach. I think we'll want to switch gears here to writing the table block as a single full parquet file.

Comment (ostafen) [reply block_file.go:12]: Okay, we could switch again to the single block per file layout, even Create()/Open() a file is bit expensive

Comment (thorfour) [reply block_file.go:12]: In the Parquet file format the file metadata is contained in the footer, we'd have to append row groups until we closed out the file with the footer. Which means in the event of a crash we'd either have to throw all the row groups away, or we'd have to do some sort of file repair, and we couldn't easily read back the row groups without the file metadata. So for the sake of complexity (with the potential tradeoff on performance as you've rightly mentioned) I think we should start with dropping full Parquet files. We can always look into merging the files out of band later.

Comment (thorfour) [reply block_file.go:12]: Yea we'll need to have some compaction settings we can tweak but I think that compaction should be outside of the scope of this work to keep the complexity low.

We can always play around with memory mapping the files as well.

If we give the parquet filenames with a ULID we should be able use that to perform timestamp comparison no?

Comment (ostafen) [reply block_file.go:12]: > If we give the parquet filenames with a ULID we should be able use that to perform timestamp comparison no?

If we get rid of compaction for now, then yes, but we will have to handle this when dealing with compacted files, because they contain multiple blocks. For now, I think it is enought to use "collectionName:timestamp" as a filename.

**Comment by ostafen** [comment]
@thorfour, each block is now stored on a separate parquet file. I also added to methods `WithPersistency()` and `WithStoragePath()` to allow the user enable persistency and to change the root directory of the store.
Actually, files are arranged in the following way: starting from the root storage path, each database has its own directory, inside of which we find separate directories for each table.

**Comment by thorfour** [comment]
> 

So awesome! Just so you know Kubecon EU is about to get into full swing here so I may not be able to get to reviewing this until next week.

**Comment by brancz** [comment]
Looooove the direction of this!

One thing that I think we've wanted to keep in mind when working on this, is that we would like object storage to be first class citizen for storage, so I'm curious if we should use [objstore](https://github.com/thanos-io/objstore) right away, so we don't use `os.File` right from the start? I was talking to the folks at segment that work on the parquet-go library, and we felt in that a file cache fill become important (that is not mmap), and they ended up writing it: https://pkg.go.dev/github.com/segmentio/datastructures/v2/pagecache

**Comment by ostafen** [comment]
So basically, should each table be backed by a different objstore.Bucket object? Or should we use a single Bucket for each database?

**Comment by ostafen** [comment]
I wrote a first version using **objstore**. Naturally, because of this abstraction, I had to remove the in memory map storing the references to open files

**Review by thorfour — COMMENTED** [review]
I think this approach is good! There's some cleanup that needs to happen before this PR is ready to merge but I think this is great!

Of note I think we'd like to change the filename of the block files to use a ULID

**Comment by ostafen** [comment]
What do you say of using this module for the ULID: https://github.com/oklog/ulid?

**Comment by brancz** [comment]
Yeah that's the one we've used in other places successfully.

**Comment by ostafen** [comment]
Used **ULID** for block filenames. I think that we should add a `Close()` method to the Table struct in order to save the current active block before exiting, or it will be always be lost (in the case the activeMemorySize is large enougth that block rotations never happens, we will not save anything on disk).

**Comment by brancz** [comment]
Something we've done in the Thanos project that worked out well is to put the actual data into a path that is just prefixed with the ULID, so something like:

```
/<arbitrary-configurable-prefix>/<database>/<table>/<ULID>/data.parquet
```

As that way we can more easily colocate files that logically belong to that block of data, for example for when we add additional metadata or indexes, they can then be colocated as:

```
/<arbitrary-configurable-prefix>/<database>/<table>/<ULID>/meta.json
/<arbitrary-configurable-prefix>/<database>/<table>/<ULID>/index
```

It allows us to be a more extensible for future additions.

File: db.go
Comment (brancz) [inline db.go:144]: Can we maybe just encapsulate the prefix using a wrapper that implements the `Bucket` interface? Then the table doesn't even need to know anything about the prefix(es). I just checked and unfortunately the objstore package doesn't seem to have that yet, but I'd be happy implementing that wrapper within this repo for now.

**Review by brancz — COMMENTED** [review]
Assuming that we can run arcticdb without issues with this feature disabled while we work out compaction etc. by just skipping the option that enables this, I think I'd be happy with merging this after this round of comments, but would also like @thorfour's opinion.

In case it wasn't clear yet, this work is absolutely incredible, thank you so much for tackling this!!

File: block_file.go
Comment (ostafen) [reply block_file.go:1]: Yes, I forgot to delete it, sorry :=)

File: db.go
Comment (ostafen) [reply db.go:130]: the same goes for this

File: db.go
Comment (ostafen) [reply db.go:144]: Yes, it seems a good idea. I'll add a BucketDecorator object

File: table_test.go
Comment (thorfour) [inline table_test.go:55]: I noticed that if I remove this, I start getting a test failure.

```
--- FAIL: Test_Table_InsertCancellation (0.06s)
    --- FAIL: Test_Table_InsertCancellation/1024 (0.06s)
        table_test.go:1115:
            	Error Trace:	table_test.go:1115
            	Error:      	"400" is not less than "400"
            	Test:       	Test_Table_InsertCancellation/1024
FAIL
exit status 1
FAIL	github.com/polarsignals/arcticdb	4.775s
```

**Review by thorfour — COMMENTED** [review]
This is awesome! Thanks so much for all the hard work you've put into this major feature. 
I agreee with @brancz I think the design is ready to go, we just need to go through some final review.

File: table_test.go
Comment (ostafen) [reply table_test.go:55]: Very strange, I just run the following command (with and without the bucket) and it works.

```bash
go test -v -run Test_Table_InsertCancellation -count 100
```

File: store.go
Comment (ostafen) [reply store.go:18]: @brancz suggested to put each block in a folder whose name is the ULID:

`/<arbitrary-configurable-prefix>/<database>/<table>/<ULID>/data.parquet`

Do you want to also repeat the ULID in the filename of the block?

File: store.go
Comment (ostafen) [reply store.go:73]: @brancz suggested to put each block in a folder whose name is the ULID:

`/<arbitrary-configurable-prefix>/<database>/<table>/<ULID>/data.parquet`

Do you want to also repeat the ULID in the filename of the block?

File: store.go
Comment (ostafen) [reply store.go:18]: What do you say if we call this `Persist()` or maybe `Upload()` (since the bucket interface contains the `Upload()` method)?

File: store.go
Comment (ostafen) [reply store.go:28]: Should we also remove the `DynamicCloserRowGroup` from the dynparquet/schema.go file? It would be no longer used by removing MemDynamicRowGroup

Comment (thorfour) [reply store.go:28]: Yes we should, thanks

File: store.go
Comment (thorfour) [reply store.go:73]: Ah ok, I think I glossed over that. It's good as is then. Ignore this comment

**Review by thorfour — APPROVED** [review]
I think we'll likely find room for optimizations here but I think this is a great foundation!

**Comment by ostafen** [comment]
I solved the conflicts with the main branch caused by recent PRs and also implemented the Bucket wrapper suggested by @brancz :=)

**Comment by brancz** [comment]
Apart from the linter job, I think this is ready to merge!

File: db.go
Comment (brancz) [inline db.go:10]: the linter wants this to be in a separate block as it's an "internal" import rather than an external dependency

File: store.go
Comment (brancz) [inline store.go:11]: same with this one

File: table.go
Comment (brancz) [inline table.go:935]: the linter wants this variable to be named `t` as that's what it's called everywhere else

**Comment by ostafen** [comment]
Should be ok now :=)

File: db.go
Comment (brancz) [inline db.go:15]: I'm so sorry, I phrased myself poorly, the way the linter wants it is to have one import statement where stdlib, external and internal imports are blocks that are separated just by newlines, so what it wants here is:

```
import (
	"fmt"
	"path"
	"sync"
	"time"

	"github.com/go-kit/log"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/thanos-io/objstore"
	"go.uber.org/atomic"

	"github.com/polarsignals/arcticdb/query/logicalplan"
)
```

**Comment by ostafen** [comment]
Should be ok now :=)

**Comment by brancz** [comment]
Amazing!! Thank you so much!!

**Pull request merged**
